### ISSUE TITLE
Cannot get expected size when InsufficientBuffer error in transmit function

### ISSUE THREAD

**Issue by ndusart** [bluetech/pcsc-rust#24]
Hello @bluetech,

From the rust api, we lose the possibility to know the required receive buffer length for call to `SCardTransmit`. We have some cards that behaves strangely and respond with data even when no data is asked in the APDU.

Normally, from the C API, the variable pointed by `pcbRecvLength` is modified to hold the expected size in case `SCARD_E_INSUFFICIENT_BUFFER` is returned. It allow us to redo the same command with appropriate buffer length.

Would it be possible to have an alternative signature for `Card::transmit` which would allow to receive the expected size back ?

Something like that ?

```
fn transmit2<'buf>(&self, send_buffer: &[u8], receive_buffer: &'buf mut[u8], expected_size: &mut usize) -> Result<&'buf [u8], Error>
```

A good fit could also be something like `status2_owned` letting the responsability of the function to allocate memory itself. Maybe something like that:

```
fn transmit_owned(&self, send_buffer: &[u8], receive_size_hint: usize) -> Result<Vec<u8>, Error>
```

If `SCARD_E_INSUFFICIENT_BUFFER` is received, the function could retry by allocating the expected size. `receive_size_hint` is the size used for the first try. If `SCARD_E_INSUFFICIENT_BUFFER` is kept being sent back, then return `Error::InsufficientBuffer`.

I can work on a PR if any of these propositions suits you.

Kind Regards,

**Comment by bluetech** [comment]
Hi @ndusart,

I think we should make the information available, it is really an oversight.

Regarding `transmit2`, I'm not sure I like `expected_size: &mut usize`, in Rust out params are not common because it's possible to have multiple return values. So it should be `-> Result<&'buf [u8], (Error, usize)>` instead, but that is less ergonomic. Ideally we could enrich the error itself with the info somehow instead of using a tuple, e.g. change `Error:InsufficientBuffer` to `Error::InsufficientBuffer(usize)`. But that would be a breaking change. Maybe some other solution?

Regarding `transmit_owned` - if transmit fails with `SCARD_E_INSUFFICIENT_BUFFER`, has the APDU actually executed (processed by the card) or does it error before actually transmitting? I'm pretty sure it's the former, since the R-APDU can contain more bytes than the `Le`. So I think that `transmit_owned` with retry is not something to encourage as it may end up executing the same APDU more than once.

**Comment by ndusart** [comment]
I agree with your concern, the expected size should be part of the result type. The thing is that `SCARD_E_INSUFFICIENT_BUFFER` is not always attached to a "correct" length (seems to be only the case for `SCardTransmit`) so it can be weird to attach it to `Error::InsufficientBuffer`. Or maybe with an `Option<usize>` ? 

And yes, the APDU will already have been executed by the card when we receive the `SCARD_E_INSUFFICIENT_BUFFER`. It is the response of the card that allows pcsclite to indicate the required size. So yes, actually it will end up executing multiple APDU. I can effectively see unexpected side-effects if this function is preferred over the other by a dev and leads to duplicate commands which are not idempotent. So I share your concern on this one too.

For me, an additional signature returning `-> Result<&'buf [u8], (Error, usize)>`  would be more than enough and it won't break existing code.
I'm afraid I cannot think of a more ergonomic solution, as you mentioned :/

**Comment by bluetech** [comment]
OK, let's do `transmit2` then. A better idea might come to mind, but I don't think so. In the documentation of this function, should just say "like `transmit()`, but ..." and explain the use case, and also that one needs to be careful about retrying non-idempotent commands.
